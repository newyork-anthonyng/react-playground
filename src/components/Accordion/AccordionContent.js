import React from "react";
import T from "prop-types";

const AccordionContent = ({ id, ariaLabelledBy, isOpen, children }) => {
  if (!isOpen) return null;

  return (
    <dd role="region" aria-labelledby={ariaLabelledBy}>
      {children}
    </dd>
  );
};

AccordionContent.defaultProps = {
  isOpen: false
};

AccordionContent.propTypes = {
  ariaLabelledBy: T.string,
  id: T.string,
  isOpen: T.bool,
  children: T.node
};

export default AccordionContent;
