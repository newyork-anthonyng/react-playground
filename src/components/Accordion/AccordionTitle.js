import React from 'react';
import T from 'prop-types';
import noop from '../../utility/noop';

const AccordionTitle = ({
  children,
  isOpen,
  ariaControls,
  onClick,
  id,
}) => (
  <dt role="heading" aria-level="3">
    <button
      aria-expanded={`${isOpen}`}
      aria-controls={ariaControls}
      onClick={onClick}
      id={id}
    >
      {children}
    </button>
  </dt>
);

AccordionTitle.defaultProps = {
  isOpen: false,
  onClick: noop,
};

AccordionTitle.propTypes = {
  children: T.node,
  isOpen: T.bool,
  ariaControls: T.string,
  id: T.string,
  onClick: T.func,
};

export default AccordionTitle;
