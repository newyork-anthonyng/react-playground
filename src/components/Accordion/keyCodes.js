export const UP = 38;
export const DOWN = 40;
export const HOME = 36;
export const END = 35;
export const ENTER = 13;
export const SPACE = 32;
