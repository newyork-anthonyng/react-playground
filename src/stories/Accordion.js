import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, number, boolean, text } from '@storybook/addon-knobs';

import {
  AccordionContainer,
  Accordion,
  AccordionTitle,
  AccordionContent,
} from '../components/Accordion';

storiesOf('Accordion', module)
  .addDecorator(withKnobs)
  .add('Example', () => {
    return (
      <AccordionContainer>
        <Accordion>
          <AccordionTitle>First</AccordionTitle>
          <AccordionContent>
            <h1>I won Gold</h1>
          </AccordionContent>
        </Accordion>
        <Accordion>
          <AccordionTitle>Second</AccordionTitle>
          <AccordionContent>
            <h1>I won Silver</h1>
            <fieldset>
              <label>
                First Name
                <input type="text" placeholder="First name" />
              </label>
              <label>
                Last Name
                <input type="text" placeholder="Last name" />
              </label>
            </fieldset>
          </AccordionContent>
        </Accordion>
        <Accordion>
          <AccordionTitle>Third</AccordionTitle>
          <AccordionContent>
            <h1>I won Bronze</h1>
          </AccordionContent>
        </Accordion>
      </AccordionContainer>
    );
  });
