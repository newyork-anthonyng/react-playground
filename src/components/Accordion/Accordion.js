import React, { Component } from 'react';
import T from 'prop-types';
import AccordionContent from './AccordionContent';
import AccordionTitle from './AccordionTitle';
import uuid from '../../utility/uuid';

class Accordion extends Component {
  constructor() {
    super();

    this.uniqueId = uuid();
  }

  renderAccordionContent = () => {
    const {
      isOpen,
      onClick,
      onKeyDown,
    } = this.props;

    const contentId = `${this.uniqueId}-content`;
    const titleId = `${this.uniqueId}-title`;

    return React.Children.map(this.props.children, (child) => {
      if (child.type === AccordionTitle) {
        return React.cloneElement(
          child,
          {
            isOpen,
            onClick,
            id: titleId,
            ariaControls: contentId,
            onKeyDown: onKeyDown,
            buttonRef: (button) => this.button = button,
          }
        );
      } else if (child.type === AccordionContent) {
        return React.cloneElement(
          child,
          {
            id: contentId,
            ariaLabelledby: titleId,
            isOpen,
          }
        );
      } else {
        throw new Error(`Accordion expected children to be AccordionTitle or AccordionContent, but received ${child.type}`);
      }
    });
  }

  render() {
    return (
      <div>
        {this.renderAccordionContent()}
      </div>
    );
  }
}

Accordion.propTypes = {
  children: T.node,
  onClick: T.func,
  isOpen: T.bool,
  onKeyDown: T.func,
};

export default Accordion;
