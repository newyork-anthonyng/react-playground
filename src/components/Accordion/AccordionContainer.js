import React, { Component } from 'react';
import * as KeyCodes from './keyCodes';
import T from 'prop-types';


class AccordionContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      buttons: React.Children.map(props.children, child => false)
    };
  }

  handleKeyDown = index => {
    return e => {
      switch (e.keyCode) {
        case KeyCodes.UP:
          this.scrollUp(index);
          break;
        case KeyCodes.DOWN:
          this.scrollDown(index);
          break;
        case KeyCodes.HOME:
          this.refs[`accordion-child-0`].button.focus();
          break;
        case KeyCodes.END:
          this.refs[
              `accordion-child-${React.Children.count(this.props.children) - 1}`
          ].button.focus();
          break;
        case KeyCodes.ENTER:
        case KeyCodes.SPACE:
          e.preventDefault();
          this.toggleContent(index);
          break;
      }
    };
  };

  handleClick = index => {
    return _ => {
      this.toggleContent(index);
    };
  };

  scrollUp = index => {
    let previousIndex = index - 1;
    if (previousIndex < 0)
    previousIndex = React.Children.count(this.props.children) - 1;

    this.refs[`accordion-child-${previousIndex}`].button.focus();
  };

  scrollDown = index => {
    let nextIndex = index + 1;
    if (nextIndex >= React.Children.count(this.props.children)) nextIndex = 0;

    this.refs[`accordion-child-${nextIndex}`].button.focus();
  };

  toggleContent = index => {
    this.setState({
      buttons: this.state.buttons.map((button, i) => {
        if (index === i) return !button;
        return button;
      })
    });
  };

  render() {
    return (
      <div>
        {React.Children.map(this.props.children, (child, i) => {
          return React.cloneElement(child, {
            onKeyDown: this.handleKeyDown(i),
            onClick: this.handleClick(i),
            ref: `accordion-child-${i}`,
            isOpen: this.state.buttons[i]
          });
        })}
      </div>
    );
  }
}

AccordionContainer.propTypes = {
  children: T.node.isRequired,
};

export default AccordionContainer;
