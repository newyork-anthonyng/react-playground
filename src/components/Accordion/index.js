import Accordion from "./Accordion";
import AccordionContainer from "./AccordionContainer";
import AccordionContent from "./AccordionContent";
import AccordionTitle from "./AccordionTitle";

export { Accordion, AccordionContainer, AccordionContent, AccordionTitle };
