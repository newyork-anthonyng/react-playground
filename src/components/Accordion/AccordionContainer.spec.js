import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import AccordionContainer from "./AccordionContainer";
import Accordion from "./Accordion";
import AccordionTitle from "./AccordionTitle";
import AccordionContent from "./AccordionContent";
import * as KeyCodes from "./keyCodes";
import noop from "../../utility/noop";
jest.mock("../../utility/uuid");

it("should render correctly", () => {
  const wrapper = shallow(
    <AccordionContainer>
      <Accordion>
        <AccordionTitle>First</AccordionTitle>
        <AccordionContent>Gold Medal</AccordionContent>
      </Accordion>
    </AccordionContainer>
  );

  expect(toJSON(wrapper)).toMatchSnapshot();
});

describe("Opening Accordion", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <AccordionContainer>
        <Accordion>
          <AccordionTitle>First</AccordionTitle>
          <AccordionContent>Gold Medal</AccordionContent>
        </Accordion>
      </AccordionContainer>
    );
  });

  it("should open Accordion when clicking on title", () => {
    const title = wrapper.find(AccordionTitle);
    title.props().onClick();

    expect(wrapper.find(Accordion).props().isOpen).toEqual(true);
  });

  it("should open Accordion when pressing ENTER", () => {
    const title = wrapper.find(AccordionTitle);
    const event = {
      preventDefault: noop,
      keyCode: KeyCodes.ENTER
    };
    title.props().onKeyDown(event);

    expect(wrapper.find(Accordion).props().isOpen).toEqual(true);
  });

  it("should open Accordion when pressing SPACE", () => {
    const title = wrapper.find(AccordionTitle);
    const event = {
      preventDefault: noop,
      keyCode: KeyCodes.SPACE
    };
    title.props().onKeyDown(event);

    expect(wrapper.find(Accordion).props().isOpen).toEqual(true);
  });
});

describe("Closing Accordion", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <AccordionContainer>
        <Accordion>
          <AccordionTitle>First</AccordionTitle>
          <AccordionContent>Gold Medal</AccordionContent>
        </Accordion>
      </AccordionContainer>
    );
    wrapper.setState({
      buttons: [true]
    });
  });

  it("should close Accordion when clicking on title", () => {
    const title = wrapper.find(AccordionTitle);
    title.props().onClick();

    expect(wrapper.find(Accordion).props().isOpen).toEqual(false);
  });

  it("should close Accordion when pressing ENTER", () => {
    const title = wrapper.find(AccordionTitle);
    const event = {
      preventDefault: noop,
      keyCode: KeyCodes.ENTER
    };
    title.props().onKeyDown(event);

    expect(wrapper.find(Accordion).props().isOpen).toEqual(false);
  });

  it("should close Accordion when pressing SPACE", () => {
    const title = wrapper.find(AccordionTitle);
    const event = {
      preventDefault: noop,
      keyCode: KeyCodes.SPACE
    };
    title.props().onKeyDown(event);

    expect(wrapper.find(Accordion).props().isOpen).toEqual(false);
  });
});

describe("When pressing UP key", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <AccordionContainer>
        <Accordion>
          <AccordionTitle>First</AccordionTitle>
          <AccordionContent>Gold Medal</AccordionContent>
        </Accordion>
        <Accordion>
          <AccordionTitle>Second</AccordionTitle>
          <AccordionContent>Silver Medal</AccordionContent>
        </Accordion>
        <Accordion>
          <AccordionTitle>Third</AccordionTitle>
          <AccordionContent>Bronze Medal</AccordionContent>
        </Accordion>
      </AccordionContainer>
    );
  });

  it("should go to previous Accordion Title", () => {
    const secondAccordionTitleEl = wrapper
      .find(AccordionTitle)
      .at(1)
      .find("button")
      .get(0);
    secondAccordionTitleEl.focus();
    expect(document.activeElement).toEqual(secondAccordionTitleEl);

    const secondAccordionTitle = wrapper.find(AccordionTitle).at(1);
    secondAccordionTitle.props().onKeyDown({ keyCode: KeyCodes.UP });

    expect(document.activeElement).toEqual(
      wrapper.find(AccordionTitle).at(0).find("button").get(0)
    );
  });

  it("should go to last Accordion Title, if focused on the first one", () => {
    const firstAccordionTitleEl = wrapper
      .find(AccordionTitle)
      .at(0)
      .find("button")
      .get(0);
    firstAccordionTitleEl.focus();
    expect(document.activeElement).toEqual(firstAccordionTitleEl);

    const firstAccordionTitle = wrapper.find(AccordionTitle).at(0);
    firstAccordionTitle.props().onKeyDown({ keyCode: KeyCodes.UP });

    expect(document.activeElement).toEqual(
      wrapper.find(AccordionTitle).at(2).find("button").get(0)
    );
  });
});

describe("When pressing DOWN key", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <AccordionContainer>
        <Accordion>
          <AccordionTitle>First</AccordionTitle>
          <AccordionContent>Gold Medal</AccordionContent>
        </Accordion>
        <Accordion>
          <AccordionTitle>Second</AccordionTitle>
          <AccordionContent>Silver Medal</AccordionContent>
        </Accordion>
        <Accordion>
          <AccordionTitle>Third</AccordionTitle>
          <AccordionContent>Bronze Medal</AccordionContent>
        </Accordion>
      </AccordionContainer>
    );
  });

  it("should go to next Accordion Title", () => {
    const secondAccordionTitleEl = wrapper
      .find(AccordionTitle)
      .at(1)
      .find("button")
      .get(0);
    secondAccordionTitleEl.focus();
    expect(document.activeElement).toEqual(secondAccordionTitleEl);

    const secondAccordionTitle = wrapper.find(AccordionTitle).at(1);
    secondAccordionTitle.props().onKeyDown({ keyCode: KeyCodes.DOWN });

    expect(document.activeElement).toEqual(
      wrapper.find(AccordionTitle).at(2).find("button").get(0)
    );
  });

  it("should go to first Accordion Title, if focused on the last one", () => {
    const thirdAccordionTitleEl = wrapper
      .find(AccordionTitle)
      .at(2)
      .find("button")
      .get(0);
    thirdAccordionTitleEl.focus();
    expect(document.activeElement).toEqual(thirdAccordionTitleEl);

    const thirdAccordionTitle = wrapper.find(AccordionTitle).at(2);
    thirdAccordionTitle.props().onKeyDown({ keyCode: KeyCodes.DOWN });

    expect(document.activeElement).toEqual(
      wrapper.find(AccordionTitle).at(0).find("button").get(0)
    );
  });
});

describe("HOME/END keys", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <AccordionContainer>
        <Accordion>
          <AccordionTitle>First</AccordionTitle>
          <AccordionContent>Gold Medal</AccordionContent>
        </Accordion>
        <Accordion>
          <AccordionTitle>Second</AccordionTitle>
          <AccordionContent>Silver Medal</AccordionContent>
        </Accordion>
        <Accordion>
          <AccordionTitle>Third</AccordionTitle>
          <AccordionContent>Bronze Medal</AccordionContent>
        </Accordion>
      </AccordionContainer>
    );
  });

  it("should go to first Accordion Title when pressing HOME key", () => {
    const secondAccordionTitleEl = wrapper
      .find(AccordionTitle)
      .at(1)
      .find("button")
      .get(0);
    secondAccordionTitleEl.focus();
    expect(document.activeElement).toEqual(secondAccordionTitleEl);

    const secondAccordionTitle = wrapper.find(AccordionTitle).at(1);
    secondAccordionTitle.props().onKeyDown({ keyCode: KeyCodes.HOME });

    expect(document.activeElement).toEqual(
      wrapper.find(AccordionTitle).at(0).find("button").get(0)
    );
  });

  it("should go to last Accordion Title when pressing END key", () => {
    const secondAccordionTitleEl = wrapper
      .find(AccordionTitle)
      .at(1)
      .find("button")
      .get(0);
    secondAccordionTitleEl.focus();
    expect(document.activeElement).toEqual(secondAccordionTitleEl);

    const secondAccordionTitle = wrapper.find(AccordionTitle).at(1);
    secondAccordionTitle.props().onKeyDown({ keyCode: KeyCodes.END });

    expect(document.activeElement).toEqual(
      wrapper.find(AccordionTitle).at(2).find("button").get(0)
    );
  });
});
