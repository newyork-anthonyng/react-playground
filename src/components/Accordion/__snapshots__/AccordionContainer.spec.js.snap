// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <Accordion
    isOpen={false}
    onClick={[Function]}
    onKeyDown={[Function]}
  >
    <AccordionTitle
      isOpen={false}
      onClick={[Function]}
    >
      First
    </AccordionTitle>
    <AccordionContent
      isOpen={false}
    >
      Gold Medal
    </AccordionContent>
  </Accordion>
</div>
`;
